File: csharp_dotnetcore/51.Enterprise-Bot-Template/Dialogs/Escalate/EscalateDialog.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder.Dialogs;

namespace EnterpriseBot
{
    public class EscalateDialog : EnterpriseDialog
    {
        // Constants
        public const string Name = "EscalateDialog";

        // Fields
        private EscalateResponses _responder = new EscalateResponses();

        public EscalateDialog(BotServices botServices)
            : base(botServices, Name)
        {
            var escalate = new WaterfallStep[]
            {
                SendPhone,
            };

            AddDialog(new WaterfallDialog(Name, escalate));
        }

        private async Task<DialogTurnResult> SendPhone(DialogContext dc, WaterfallStepContext step, CancellationToken cancellationToken)
        {
            await _responder.ReplyWith(dc.Context, EscalateResponses.SendPhone);
            return await dc.EndAsync();
        }
    }
}

File: samples/javascript_typescript/52.enterprise-bot/src/dialogs/escalate/escalateDialog.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { DialogTurnResult, WaterfallDialog, WaterfallStepContext } from "botbuilder-dialogs";
import { BotServices } from "../../botServices";
import { EnterpriseDialog } from "../shared/enterpriseDialog";
import { EscalateResponses } from "./escalateResponses";

export class EscalateDialog extends EnterpriseDialog {
    // Fields
    public static readonly _responder: EscalateResponses = new EscalateResponses();

    private static async SendPhone(sc: WaterfallStepContext): Promise<DialogTurnResult> {
        await EscalateDialog._responder.replyWith(sc.context, EscalateResponses.SendPhone);
        return await sc.endDialog();
    }

    constructor(botServices: BotServices) {
        super(botServices, EscalateDialog.name);
        this.initialDialogId = EscalateDialog.name;

        const escalate = [
            EscalateDialog.SendPhone.bind(this),
        ];

        this.addDialog(new WaterfallDialog(this.initialDialogId, escalate));
    }
}
